### ISSUE TITLE
Support ring async?

### ISSUE THREAD

**Issue by vincentjames501** [http-kit/http-kit#394]
It would be awesome if http-kit server could support the ring async handler mechanism.

```clojure
(defn handler
  ([request]
   {:status 200, :headers {}, :body "Hello World"})
  ([request respond raise]
   (respond {:status 200, :headers {}, :body "Hello World"}))
```

Right now it appears to always use the synchronous mechanism:

```java
Map resp = (Map)this.handler.invoke(ClojureRing.buildRequestMap(this.req));
```

I haven't tested, but I suspect this could be tweaked and if `this.handler` is of type `MultiFn`, it could invoke the 3 arity handler instead and the `HttpHandler` thread would have to exit so it can be truly async and not block a request thread.

**Comment by vemv** [comment]
Seconded!

`MultiFn` refers to multimethods rather than multi-arity functions though, so the suggestion may be off.

**Comment by vincentjames501** [comment]
Yeah I would probably just do what the jetty version did and just specify an option to the server and it will invoke the three arg handler instead. Easy enough.

**Comment by vemv** [comment]
Right, https://github.com/ring-clojure/ring/blob/6a1c7a17f9d1979ac00e3cb689ba0c1f61fe716b/ring-jetty-adapter/src/ring/adapter/jetty.clj#L35

One disadvantage with `ring.adapter.jetty` seems to be that the choice is coarse-grained, i.e. either all of none of your handlers will have the 3-arg signature.

Perhaps if one passed around vars rather than fns (i.e. `#'my-handler` vs. `my-handler`), the http-kit adapter could inspect (`(-> #'my-handler :meta :arglists)`), determine the existing signature, and invoke that one.

The performance impact would be negligible AFAICT (clojure introspection is cheap; invoking a var instead of its underlying fn as a small fixed cost).

**Comment by weavejester** [comment]
> One disadvantage with `ring.adapter.jetty` seems to be that the choice is coarse-grained, i.e. either all of none of your handlers will have the 3-arg signature.

How does HTTP-Kit support multiple handlers? I thought you just passed one handler to `run-server`.

**Comment by vincentjames501** [comment]
I think my original thought should be ignored. I think there should simply be a new config param like `ring.adapter.jetty` and pass `:async? true` which would then invoke a 3 arg handler instead.

**Comment by vincentjames501** [comment]
@weavejester , I think @vemv is just implying that you can't have a "middleware chain" where some are 1 arg handlers and others are 3. We ran into that as well when we converted to using jetty async and several 3rd party middlewares we had to either abandon or update to add the 3 arg variant.

**Comment by jconti** [comment]
I think the principal benefit of having a ring-compatible adapter is to reuse the Ring ecosystem as-is.  Since it appears that the dispatch to handlers require all asynchronous or all synchronous signatures and are principally designed [*not* to support long-polling and websocket](https://groups.google.com/d/msg/ring-clojure/JD9FLJFTVsg/xhm7fUWU6asJ) as the http-kit unified async API is, emulating the Jetty style seems to meet the requirement.  But it would preclude being able to use some of the features of http-kit.  I think that is a reasonable tradeoff, reuse ecosystem software, or, get the full power of the api.  Documentation to make it clear would be very desirable.  A routing library like [Reitit](https://github.com/metosin/reitit) and others may permit better solutions while still supporting ring on some routes and exposing the unified http-kit api on others.

**Comment by ptaoussanis** [comment]
Support will be available in the upcoming `v2.8.0-beta2`, courtesy of @weavejester 🚀🙏
